src/Avalonia.Base/PropertyStore/LocalValueBindingObserver.cs

```diff
@@ -74,11 +78,21 @@ static void Execute(LocalValueBindingObserver<T> instance, BindingValue<T> value
                 LoggingUtils.LogIfNecessary(owner.Owner, property, value);
 
                 if (value.HasValue)
-                    owner.SetValue(property, value.Value, BindingPriority.LocalValue);
-                else if (value.Type != BindingValueType.DataValidationError)
-                    owner.ClearLocalValue(property);
+                {
+                    var effectiveValue = value.Value;
+                    if (property.ValidateValue?.Invoke(effectiveValue) == false)
```

Comment:
Fixes a bug here where we actually weren't validating the value in this case. Also added a unit test.